qutip/qobj.py

```diff
@@ -328,6 +328,13 @@ def __init__(self, inpt=None, dims=None, shape=None,
             if self.type == 'super' and self.superrep is None:
                 self.superrep = 'super'
 
+        if (self._data.shape[0] > np.prod(np.hstack(self.dims[0])) or
+           self._data.shape[1] > np.prod(np.hstack(self.dims[1]))) and \
```

Comment:
I'm a bit surprised that the check is whether the shape `>` to the product of the dims, rather than simply `!=`. Is there a situation where being `<` the dims is okay?

Reply:
Ericgig pointed out that `enr_` functions return data where `<` is OK. The `>` check allows these cases. This does however leave open the possibility of misshapen data where this condition is technically fulfilled i.e. `Qobj(complex[n,1],dims=[[n],[n]])`. This seems however a very niche and possibly from the start an erroneous case.

Reply:
We should definitely at least at a comment about this here.

I am wondering if we can do something better (especially in QuTiP 5).

@Ericgig Any thoughts?

Reply:
This is something we definitively need to take into account when making the dims class in v5 and a reason we need to do it sooner than later. For v4, a comment could be enough. Or we could have the constructor being strict and then 3~4 special functions to work around it.

Reply:
@Lala5th Could you add the comment here and then I'm happy to merge this?
@Ericgig If you could approve too, that would be good (since the QObj constructor is one of the most used parts of the code)

Reply:
@hodgestar Is this what you had in mind? Some checks (2) unfortunately should still fail, but I couldn't find proper documentation on those to figure out if they are valid use cases or some ill formed code that has not been noticed until now. (The two checks are the ones relating to `TestHierarchyADOsState` in `test_bofin_solvers.py`)

Reply:
That looks great -- thanks!

Reply:
I can look at the HEOM / BoFiN test failures. Quite possibly I just got some dims or shapes wrong in those tests.